### PR TITLE
Implement and test S3 Sink, with usage in runtime.

### PR DESCRIPTION
**Pull request by aliddell**

### CODE REVIEW

**Review by shlomnissan — CHANGES_REQUESTED** [review]
I don’t have clear blockers, but I think a round of cleanup and clarifications will be good before merging. There are several concerns that I have in general:

- I'm worried that the error handling patterns in this feature are unclear. Using catch-all for exceptions and the practice of throwing for catching and logging is especially concerning.
- I think we're using `noexcept` inappropriately (primarily to convey intent), which could cause issues over time.
- I might be missing some context, but the use of certain containers (`std::list` and `std::queue`) doesn’t make a lot of sense to me. These are not trivial decisions because they typically convey intent and come with overhead.

Anyway, I think it’s almost there.

File: src/writers/file.sink.cpp
Comment (aliddell) [reply src/writers/file.sink.cpp:19]: I'm slow-rolling this replacement in another branch.

File: src/common/utilities.hh
Comment (aliddell) [reply src/common/utilities.hh:27]: There is a `static_cast` [here](https://github.com/acquire-project/acquire-driver-zarr/blob/238-pt-6/src/writers/sink.creator.cpp#L138). Starting with 2 is to eliminate any confusion that might cause.

File: src/common/utilities.cpp
Comment (aliddell) [reply src/common/utilities.cpp:204]: I think I like this better. Let me try it out.

File: src/writers/zarrv3.writer.cpp
Comment (aliddell) [reply src/writers/zarrv3.writer.cpp:33]: I think this is fine, but let's hold off on it for a separate PR.

File: src/writers/s3.sink.hh
Comment (aliddell) [reply src/writers/s3.sink.hh:9]: We are not. I was thinking about it when I remembered we were using threads anyway.

File: src/writers/s3.sink.hh
Comment (aliddell) [reply src/writers/s3.sink.hh:15]: Understood. I'm one of those who likes the documentation aspect of it, a habit I picked up from my last job. If it really bothers you I can remove it.

File: src/writers/s3.sink.hh
Comment (aliddell) [reply src/writers/s3.sink.hh:21]: Yup, [separate branch](https://github.com/acquire-project/acquire-driver-zarr/pull/278#discussion_r1684414792).

File: src/writers/s3.sink.hh
Comment (aliddell) [reply src/writers/s3.sink.hh:30]: I'm pretty sure I had used `std::array` but it was unsuitable for some reason. Let me try it again.

Comment (aliddell) [reply src/writers/s3.sink.hh:30]: OK, can't reproduce whatever problem I had. It's an array now.

File: src/writers/s3.sink.hh
Comment (aliddell) [reply src/writers/s3.sink.hh:34]: `minio::s3::CompleteMultipartUploadArgs` takes a `std::list`.

File: src/writers/s3.sink.cpp
Comment (aliddell) [reply src/writers/s3.sink.cpp:11]: It's generated in [`SinkCreator::make_s3_objects_()`](https://github.com/acquire-project/acquire-driver-zarr/blob/238-pt-6/src/writers/sink.creator.cpp#L359).

File: src/writers/s3.sink.cpp
Comment (aliddell) [reply src/writers/s3.sink.cpp:43]: `offset` is not being used because this is an override and it is needed for `FileSink::write()`.

File: src/writers/s3.sink.cpp
Comment (aliddell) [reply src/writers/s3.sink.cpp:129]: I like it.

File: src/zarr.cpp
Comment (aliddell) [reply src/zarr.cpp:373]: No, but it's slightly more performant because it doesn't need to use `strlen`.

File: src/writers/sink.creator.cpp
Comment (aliddell) [reply src/writers/sink.creator.cpp:34]: Yeah, this is a lesson I've learned since starting this project. This will take a change to the [StorageProperties](https://github.com/acquire-project/acquire-common/blob/main/acquire-core-libs/src/acquire-device-properties/device/props/storage.h#L43) struct, which I'm going to have to take care of separately.

File: src/writers/sink.creator.cpp
Comment (aliddell) [reply src/writers/sink.creator.cpp:348]: - I use the queue to construct paths, as we discussed in a previous PR. I'd have to rework that whole thing. I really don't think that a vector is the right tool for constructing paths like `/(1...k)/(1...m)/(1...n)/.../(1...z)`.
- I've been thinking that since we're not actually doing anything resource intensive, like opening a file, when we create an S3Sink, we can do this serially.

File: src/writers/s3.sink.hh
Comment (shlomnissan) [reply src/writers/s3.sink.hh:16]: All good, this is a personal preference thing 👍

File: src/writers/s3.sink.hh
Comment (shlomnissan) [inline src/writers/s3.sink.hh:33]: I would restore the comment or create a variable (`static constexpr size_t five_mb_size = 5 << 20`)  so it's obvious what this number represents.

File: src/writers/s3.sink.cpp
Comment (shlomnissan) [inline src/writers/s3.sink.cpp:118]: I think this should be a separate method. Sorry if my comment was confusing—I typically refer to accessor methods as computed properties.

File: src/writers/s3.sink.cpp
Comment (shlomnissan) [inline src/writers/s3.sink.cpp:42]: A common pattern for identifying unused parameters is using underscores.

```cpp
zarr::S3Sink::write(size_t _, const uint8_t* data, size_t bytes_of_data)
```

File: src/writers/sink.creator.cpp
Comment (shlomnissan) [reply src/writers/sink.creator.cpp]: I think I’m missing something here. I’ll tinker with the code to better understand this requirement.

**Review by shlomnissan — APPROVED** [review]
A couple of minor comments, otherwise, LGTM :+1:

**Pull request merged**
